docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+of the custom metric API, but also of the external metric API, preventing the deployment of alternative handlers for
+these API extension points. 
+
+### Element: HPA
+ShootKapi is scaled horizontally via HPA operating on the aforementioned average kube-apiserver request rate custom
+metric. Scale-in is delayed by a stabilization window. This avoids unnecessary flapping, the effects of which would
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
+
+### Element: VPA
+A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
+HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.
+
+This improves over the existing solution by having vertical scaling active the entire time and not only during certain
+periods.
+
+### Transition Strategy
+The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
+to the one hereby proposed.
+
+1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
+   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
+2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+4. New feature flag promoted and eventually removed.
+5. (Outside the scope of this proposal) Use of HVPA removed for other workloads and the preexisting HVPA flag removed.
+
+### Scaling Clusters with No Monitoring Stack
+Monitoring stack is not deployed on clusters with `testing` purpose. An alternative mode of autoscaling will be used for
+such clusters: In the short term, they will be scaled via the existing "class-based scaling + HPA" mode, which is the
+current alternative to HVPA.
```

Comment:
This relates to one of my comments above:

In addition to the `testing` purpose, it's possible to disable the shoot monitoring instances entirely via the gardenlet configuration. 
E.g., we don't use gardener's monitoring stack for shoots. Hence, we won't be able to make use of this feature.
This is another point for using a central prometheus instance per seed.

Reply:
On a related, but separate note:
Assume this GEP supported an extensibility option which enabled you to deploy your own provider at the "custom metrics" API extension  point. This provider would need to implement the necessary requests-per-second-per-kapi-pod-per-shoot metric, to replace the provider based on prometheus-metrics-adapter. Would that make the proposed scaling mode useful to you?

Reply:
>This is another point for using a central prometheus instance per seed.

I looked further into this and consulted the monitoring experts. The existing cache prometheus is the obvious candidate. However, having it authenticate against the individual ShootKapis would be a major technical complication. The alternative of adding additional (4th) prometheus instance on the seed, has its obvious downsides too. Overall, it is not clear that the benefits are worth it at this time (future changes could change the parameters of that calculation, I assume).

Reply:
This discussion is now outdated, due to adding the gardener-custom-metrics, and no longer depending on prometheus as data source.